enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,821 @@
+		add_and_resolve_dns_name(obj.spec.name)
+
+	for each resolvedName in obj.status.resolvedNames:
+		ensure_dns_name_added(resolvedName)
+	
+	if removal_of_ips_required(obj):
+		update_dns_resolver(obj)
+		return
+
+	remaining_time_till_grace_period, is_expiring = grace_period_expires_after(obj)
+
+	if is_expiring:
+		reconcile_event_after(remaining_time_till_grace_period)
+		return
+
+else if match(event.type, "delete"):
+
+	if is_wildcard(obj.spec.name) == true
+		|| (is_wildcard(obj.spec.name) == false && match_wildcard_dns_name(obj.spec.name) == false):
+		delete_dns_name(obj.spec.name)
+
+	if is_wildcard(obj.spec.name) == true:
+		for each resolvedName in obj.status.resolvedNames:
+			if resolvedName.dnsName != obj.spec.name && match_regular_dns_name(resolvedName.dnsName) == false:
+				delete_dns_name(obj.spec.name)
+```
+* The `DNSNameResolver` controller will run dns resolution of DNS names in the background which will refresh the `.status` field of the `DNSNameResolver`
+CRs upon expiration of the validity of the corresponding IP addresses associated to the resolved names. As the CoreDNS pods are configured using the upstream
+nameservers present in the node host's `/etc/resolv.conf` file, the upstream nameservers for different CoreDNS pods may serve different IP addresses (possibly
+with different TTLs) for the same DNS name. Thus, while refreshing the information of a DNS name, the DNS lookup requests are sent to a maximum of 5 randomly
+chosen CoreDNS pods to get as many different associated IP addresses for the DNS name as possible.
+* The `DNSNameResolver` controller will watch for the events related to `DNSNameResolver` CRs.
+* After it receives the create/update events, it will perform DNS lookup for each of the DNS names corresponding to the `DNSNameResolver` CRs,
+if the details of the DNS names are not already added.
+* The `DNSNameResolver` controller will store the regular DNS name matching the `.spec.name` field of a `DNSNameResolver` CR and the
+corresponding current IP addresses along with the the TTL and the last lookup time for each IP address. Using the TTL and last lookup time of all the associated
+IP addresses of a DNS name, a next lookup time for the DNS name is also stored. Based on the next lookup time, the controller will perform DNS
+lookups to get the latest IP addresses and TTL information. The DNS lookup will be intercepted by the `ocp_dnsnameresolver` plugin and will enforce an
+update of the `.status` of the CR.
+* The controller will also store the wildcard DNS name which matches the `.spec.name` field of a `DNSNameResolver` CR. If the DNS lookup is
+successful for the wildcard DNS name, then the DNS name will be stored with the corresponding current IP addresses along with the TTL and the last lookup time
+for each IP address. If the DNS lookup is not successful, then it will be retried after a default TTL (30 minutes). The `DNSNameResolver` controller will
```

Comment:
If it's 30 minutes, is that an improvement over the current 30 minutes polling time of Egress FW?